File: tests/bench/crosslang/c/u128div.c
/* Порт tests/bench/programs/U128DivBench.eat 1:1 — деление 128 бит:
 * EATLang — полный сдвиговый divrem (128 итераций) и divrem_32 по
 * лимбам, C — нативное деление __int128 (__udivti3). Вывод обязан
 * совпасть байт-в-байт при том же REPEAT. */
#include <stdint.h>
#include <stdio.h>

static const uint32_t REPEAT = 1;

typedef unsigned __int128 u128;

static uint64_t lcg(uint64_t x) {
    return (x * 31u + 7u) % 281474976710656ull;
}

static uint64_t work(uint64_t seed) {
    uint64_t x = seed + 3u;
    uint64_t acc = 0;
    for (uint32_t i = 0; i < 400u; i++) {
        x = lcg(x);
        uint64_t hi = x;
        x = lcg(x);
        uint64_t lo = x;
        x = lcg(x);
        uint64_t d = x | 1u;
        u128 a = ((u128)hi << 64) | lo;
        u128 q = a / d;
        u128 r = a % d;
        acc = acc ^ (uint64_t)q ^ (uint64_t)(q >> 64) ^ (uint64_t)r;
    }
    for (uint32_t i = 0; i < 4000u; i++) {
        x = lcg(x);
        uint64_t hi = x;
        x = lcg(x);
        uint64_t lo = x;
        x = lcg(x);
        uint64_t d32 = (x & 4294967295ull) | 1u;
        u128 a = ((u128)hi << 64) | lo;
        u128 q = a / d32;
        u128 r = a % d32;
        acc = acc ^ (uint64_t)q ^ (uint64_t)(q >> 64) ^ (uint64_t)r;
    }
    return acc % 65536u;
}

int main(void) {
    uint64_t acc = 7u;
    for (uint32_t r = 0; r < REPEAT; r++) {
        acc = work(acc);
    }
    uint32_t c = (uint32_t)acc;
    printf("checksum %u\n", c);
    return 0;
}

File: tests/bench/crosslang/rust/u128div.rs
// Порт tests/bench/programs/U128DivBench.eat 1:1 — деление 128 бит:
// EATLang — полный сдвиговый divrem (128 итераций) и divrem_32 по
// лимбам, Rust — нативное деление u128 (__udivti3). Вывод обязан
// совпасть байт-в-байт при том же REPEAT.

const REPEAT: u32 = 1;

fn lcg(x: u64) -> u64 {
    (x * 31 + 7) % 281474976710656
}

fn work(seed: u64) -> u64 {
    let mut x: u64 = seed + 3;
    let mut acc: u64 = 0;
    for _ in 0..400u32 {
        x = lcg(x);
        let hi = x;
        x = lcg(x);
        let lo = x;
        x = lcg(x);
        let d = x | 1;
        let a: u128 = ((hi as u128) << 64) | (lo as u128);
        let q = a / (d as u128);
        let r = a % (d as u128);
        acc = acc ^ (q as u64) ^ ((q >> 64) as u64) ^ (r as u64);
    }
    for _ in 0..4000u32 {
        x = lcg(x);
        let hi = x;
        x = lcg(x);
        let lo = x;
        x = lcg(x);
        let d32 = (x & 4294967295) | 1;
        let a: u128 = ((hi as u128) << 64) | (lo as u128);
        let q = a / (d32 as u128);
        let r = a % (d32 as u128);
        acc = acc ^ (q as u64) ^ ((q >> 64) as u64) ^ (r as u64);
    }
    acc % 65536
}

fn main() {
    let mut acc: u64 = 7;
    for _ in 0..REPEAT {
        acc = work(acc);
    }
    let c: u32 = acc as u32;
    println!("checksum {}", c);
}
